File: essentials/chapter_4/llamaindex/agent_app.ts
/**
 * LLM application to chat with multiple LLMs - LlamaIndex TypeScript framework implementation.
 */

import {
    BaseLLMManager,
    createLlamaIndexModel,
    interactiveCli,
    normalizeResponseText,
    printCliHelp,
} from '../../../shared/utils.ts';

type LlamaIndexContentBlock = {
    type?: string;
    text?: string;
};

type LlamaIndexResponse = {
    message?: {
        content?: string | LlamaIndexContentBlock[];
    } | unknown;
};

class LlamaIndexLLMManager extends BaseLLMManager {
    constructor() {
        super('LlamaIndex TypeScript');
    }

    async _testProvider(provider: string): Promise<void> {
        await this._createModel(this.providerModelIdentifier(provider), 0.7, 1000);
    }

    _createModel(selectedModel: string, temperature: number, maxTokens: number) {
        return createLlamaIndexModel(selectedModel, {
            temperature,
            maxTokens,
        });
    }

    _resolveProvider(provider: string): string | null {
        return this.resolveModelIdentifier(provider);
    }

    _extractText(result: LlamaIndexResponse): string {
        const message = result?.message;
        const content = message && typeof message === 'object' && 'content' in message
            ? message.content
            : undefined;
        return normalizeResponseText(content ?? message ?? result);
    }

    async askQuestion(
        topic: string,
        provider: string | null = null,
        template = '{topic}',
        maxTokens = 1000,
        temperature = 0.7,
    ) {
        const prompt = template.replace('{topic}', topic);
        const modelConfig = this.resolveModelConfig(provider);

        if (!modelConfig) {
            return {
                success: false,
                error: 'No providers available',
                provider: 'none',
                model: 'none',
                prompt,
                response: null,
            };
        }

        try {
            const model = this._createModel(modelConfig.name, temperature, maxTokens);
            const messages = [{ role: 'user', content: prompt }];
            const response = this._extractText(await model.chat({ messages }));

            return {
                success: true,
                provider: modelConfig.provider,
                model: modelConfig.model,
                modelIdentifier: modelConfig.name,
                prompt,
                response,
                temperature,
                maxTokens,
            };
        } catch (error: unknown) {
            return {
                success: false,
                provider: modelConfig.provider,
                model: modelConfig.model,
                modelIdentifier: modelConfig.name,
                prompt,
                error: error instanceof Error ? error.message : String(error),
                response: null,
                temperature,
                maxTokens,
            };
        }
    }
}

async function main() {
    const args = process.argv.slice(2);
    if (args.includes('-h') || args.includes('--help')) {
        printCliHelp(process.argv[1]);
        return;
    }

    if (args.includes('web')) {
        try {
            const { runWebServer } = await import('../../../shared/essentials/web.ts');
            await runWebServer(() => new LlamaIndexLLMManager());
        } catch (error) {
            console.error('Error: shared web API not found or Express not installed.');
            console.error('Install Express: npm install express cors');
            process.exit(1);
        }
    } else {
        const manager = new LlamaIndexLLMManager();
        await manager._checkProviders();
        await interactiveCli(manager);
    }
}

export { LlamaIndexLLMManager };

if (import.meta.url === `file://${process.argv[1]}`) {
    main().catch(console.error);
}

File: essentials/chapter_4/llamaindex/agent_app.py
"""LLM application to chat with multiple LLMs - LlamaIndex Python framework implementation."""

import os
import sys
from typing import Dict, Optional

sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "..", "..")))

if any(arg in {"-h", "--help"} for arg in sys.argv[1:]):
    from shared.utils import print_cli_help

    print_cli_help(sys.argv[0])
    sys.exit(0)


from llama_index.core.llms import ChatMessage
from shared.utils import BaseLLMManager, interactive_cli, normalize_response_text
from shared.utils import create_llamaindex_model


class LlamaIndexLLMManager(BaseLLMManager):
    """LlamaIndex implementation with reusable hooks for chapter extensions."""

    def __init__(self):
        super().__init__("LlamaIndex")

    def _test_provider(self, provider: str):
        self._create_model(self.provider_model_identifier(provider), temperature=0.7, max_tokens=1000)

    def _create_model(self, selected_model: str, temperature: float, max_tokens: int):
        return create_llamaindex_model(
            selected_model,
            temperature=temperature,
            max_tokens=max_tokens,
        )

    def _create_client(self, provider: str, temperature: float, max_tokens: int):
        return self._create_model(provider, temperature=temperature, max_tokens=max_tokens)

    def _resolve_provider(self, provider: Optional[str]):
        return self.resolve_model_identifier(provider)

    @staticmethod
    def _extract_text(result) -> str:
        content = getattr(getattr(result, "message", None), "content", None)
        return normalize_response_text(content if content is not None else result)

    def ask_question(
        self,
        topic: str,
        provider: str = None,
        template: str = "{topic}",
        max_tokens: int = 1000,
        temperature: float = 0.7,
    ) -> Dict:
        prompt = template.format(topic=topic)
        model_config = self.resolve_model_config(provider)

        if not model_config:
            return {
                "success": False,
                "error": "No providers available",
                "provider": "none",
                "model": "none",
                "prompt": prompt,
                "response": None,
            }

        try:
            model = self._create_model(model_config.name, temperature=temperature, max_tokens=max_tokens)
            messages = [ChatMessage(role="user", content=prompt)]
            result = model.chat(messages)
            response_text = self._extract_text(result)
            return {
                "success": True,
                "provider": model_config.provider,
                "model": model_config.model,
                "model_identifier": model_config.name,
                "prompt": prompt,
                "response": response_text,
                "temperature": temperature,
                "max_tokens": max_tokens,
            }
        except Exception as exc:
            return {
                "success": False,
                "provider": model_config.provider,
                "model": model_config.model,
                "model_identifier": model_config.name,
                "prompt": prompt,
                "error": str(exc),
                "response": None,
                "temperature": temperature,
                "max_tokens": max_tokens,
            }


def main():
    args = sys.argv[1:]
    if "web" in args:
        try:
            from shared.essentials.web import run_web_server

            run_web_server(lambda: LlamaIndexLLMManager())
        except ImportError:
            print("Error: shared web API not found or FastAPI not installed.")
            print("Install FastAPI: pip install fastapi uvicorn")
            sys.exit(1)
    else:
        manager = LlamaIndexLLMManager()
        interactive_cli(manager)


if __name__ == "__main__":
    main()
